library/x509_crt.c

```diff
@@ -1435,32 +1571,179 @@ int mbedtls_x509_crt_parse_path( mbedtls_x509_crt *chain, const char *path )
 }
 #endif /* MBEDTLS_FS_IO */
 
+/*
+ * OtherName ::= SEQUENCE {
+ *      type-id    OBJECT IDENTIFIER,
+ *      value      [0] EXPLICIT ANY DEFINED BY type-id }
+ *
+ * HardwareModuleName ::= SEQUENCE {
+ *                           hwType OBJECT IDENTIFIER,
+ *                           hwSerialNum OCTET STRING }
+ *
+ * NOTE: we currently only parse and use otherName of type HwModuleName,
+ * as defined in RFC 4108.
+ */
+int mbedtls_x509_get_other_name( const mbedtls_x509_sequence *subject_alt_name,
+                                 mbedtls_x509_san_othername *other_name )
+{
+    int ret;
+    size_t len;
+    unsigned char *p = subject_alt_name->buf.p;
+    const unsigned char *end = p + subject_alt_name->buf.len;
+    mbedtls_x509_buf cur_oid = { 0, 0, NULL };
+
+    if( subject_alt_name->buf.tag != ( MBEDTLS_ASN1_CONTEXT_SPECIFIC | 0 ) )
+    {
+        /*
+         * The given subject alternative name is not of type "othername".
+         */
+        return( MBEDTLS_ERR_X509_BAD_INPUT_DATA );
+    }
+
+    if( ( ret = mbedtls_asn1_get_tag( &p, end, &len,
+                                      MBEDTLS_ASN1_OID ) ) != 0 )
+        return( MBEDTLS_ERR_X509_INVALID_EXTENSIONS + ret );
+
+    cur_oid.tag = MBEDTLS_ASN1_OID;
+    cur_oid.p = p;
+    cur_oid.len = len;
+
+    /*
+     * Only HwModuleName is currently supported.
+     */
+    if( MBEDTLS_OID_CMP( MBEDTLS_OID_ON_HW_MODULE_NAME, &cur_oid ) != 0 )
```

Comment:
It seems that the SAN is actually a sequence of GeneralName. Is it allowed to have more than one "name" where one of them is a hwmodulename? If so, then we probably need to iterate through the sequence. Can a SAN contain mora than one otherName with different hw module name? If so, then we probably need to return a list of names here.

If none of the above are allowed by the standard, do we check this?

Reply:
Perhaps RFC 4108 mentions this, I cou;dn't find, but the SAN extension doesn't forbid a single `otherName` type. There could be several types of `otherName`

>It seems that the SAN is actually a sequence of GeneralName.

I don't follow you.  Is the SAN you are mentioning the subject alternative Name extension types, or the `otherName` type only? It is not necessarily a list of GeneralName, as it dependent on the type and sub-type of `otherName`

Reply:
I think the point of my comment is that the SAN is defined as follows:
```
id-ce-subjectAltName OBJECT IDENTIFIER ::=  { id-ce 17 }

   SubjectAltName ::= GeneralNames

   GeneralNames ::= SEQUENCE SIZE (1..MAX) OF GeneralName

   GeneralName ::= CHOICE {
        otherName                       [0]     OtherName,
        rfc822Name                      [1]     IA5String,
        dNSName                         [2]     IA5String,
        x400Address                     [3]     ORAddress,
        directoryName                   [4]     Name,
        ediPartyName                    [5]     EDIPartyName,
        uniformResourceIdentifier       [6]     IA5String,
        iPAddress                       [7]     OCTET STRING,
        registeredID                    [8]     OBJECT IDENTIFIER }

   OtherName ::= SEQUENCE {
        type-id    OBJECT IDENTIFIER,
        value      [0] EXPLICIT ANY DEFINED BY type-id }

   EDIPartyName ::= SEQUENCE {
        nameAssigner            [0]     DirectoryString OPTIONAL,
        partyName               [1]     DirectoryString }
```

My understanding from that code in the RFC 5280 is that there could be multiple otherName values in the same SAN. However, this function only returns one SAN and as far as I can tell it expects the otherName SAN to be at the front of the sequence provided as a parameter. Why is this the case?

Reply:
The function doesn't expect the `otherName` to be at the front. It expects the caller to first check the sequence item, and if it's `otherName` send it to this function. This function expects *a single sequence item* and not a list.
Anyway, this has changed and will be pushed later today

Reply:
Has this been changed?

Reply:
The change is that now this function is internal static function, and it receives a single `mbedtls_x509_sequence` item, of type `otherName`

Reply:
ok